src/cpp/ext/filters/census/environment_autodetect.cc

```diff
@@ -0,0 +1,421 @@
+                          std::string /* result */)>
+      callback_;
+  grpc_http_response response_;
+};
+
+// This is not a definite method to get the namespace name for GKE, but it is
+// the best we have.
+std::string GetNamespaceName() {
+  // Read the root file.
+  const char* filename =
+      "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
+  auto namespace_name = grpc_core::LoadFile(filename, false);
+  if (!namespace_name.ok()) {
+    gpr_log(GPR_DEBUG, "Reading file %s failed: %s", filename,
+            grpc_core::StatusToString(namespace_name.status()).c_str());
+    // Fallback on an environment variable
+    return grpc_core::GetEnv("NAMESPACE_NAME").value_or("");
+  }
+  return std::string(reinterpret_cast<const char*>((*namespace_name).begin()),
+                     (*namespace_name).length());
+}
+
+// Get pod name for GKE
+std::string GetPodName() {
+  auto pod_name = grpc_core::GetEnv("POD_NAME");
+  if (pod_name.has_value()) {
+    return pod_name.value();
+  }
+  return grpc_core::GetEnv("HOSTNAME").value_or("");
+}
+
+// Get container name for GKE
+std::string GetContainerName() {
+  return grpc_core::GetEnv("HOSTNAME").value_or("");
+}
+
+// Get function name for Cloud Functions
+std::string GetFunctionName() {
+  auto k_service = grpc_core::GetEnv("K_SERVICE");
+  if (k_service.has_value()) {
+    return k_service.value();
+  }
+  return grpc_core::GetEnv("FUNCTION_NAME").value_or("");
+}
+
+// Get revision name for Cloud run
+std::string GetRevisionName() {
+  return grpc_core::GetEnv("K_REVISION").value_or("");
+}
+
+// Get service name for Cloud run
+std::string GetServiceName() {
+  return grpc_core::GetEnv("K_SERVICE").value_or("");
+}
+
+// Get configuration name for Cloud run
+std::string GetConfiguratioName() {
+  return grpc_core::GetEnv("K_CONFIGURATION").value_or("");
+}
+
+// Get module ID for App Engine
+std::string GetModuleId() {
+  return grpc_core::GetEnv("GAE_SERVICE").value_or("");
+}
+
+// Get version ID for App Engine
+std::string GetVersionId() {
+  return grpc_core::GetEnv("GAE_VERSION").value_or("");
+}
+
+// Fire and forget class
+class EnvironmentAutoDetectHelper
+    : public grpc_core::InternallyRefCounted<EnvironmentAutoDetectHelper>,
+      private internal::GrpcLibrary {
+ public:
+  explicit EnvironmentAutoDetectHelper(
+      std::string project_id,
+      absl::AnyInvocable<void(EnvironmentAutoDetect::ResourceType)> on_done)
+      : InternallyRefCounted(/*trace=*/nullptr, /*initial_refcount=*/2),
+        project_id_(std::move(project_id)),
+        on_done_(std::move(on_done)) {
+    grpc_core::ExecCtx exec_ctx;
+    // TODO(yashykt): The pollset stuff should go away once the HTTP library is
+    // ported over to use EventEngine.
+    pollset_ = static_cast<grpc_pollset*>(gpr_zalloc(grpc_pollset_size()));
+    grpc_pollset_init(pollset_, &mu_poll_);
+    pollent_ = grpc_polling_entity_create_from_pollset(pollset_);
+    poller_ = grpc_core::Thread(
```

Comment:
This can be replaced by `GetDefaultEventEngine()->Run([this]{ PollLoop(); });`. 

Note that with either a thread like this or using EventEngine::Run, we don't strictly have to worry about fork support since this is in C++ code, not the core library where Python requires a clean fork story, but it might be worth adding a comment that this is not fork-safe for anyone who wants fork support in C++. If you want to make this fork-safe, we can have a short design chat.

Reply:
Suggestion accepted. Added comments to that affect when we actually care about fork support here.